clients/src/main/java/org/apache/kafka/common/record/OldLogEntry.java

```diff
@@ -0,0 +1,412 @@
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.CorruptRecordException;
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.common.utils.ByteBufferInputStream;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Iterator;
+
+import static org.apache.kafka.common.record.Records.LOG_OVERHEAD;
+import static org.apache.kafka.common.record.Records.OFFSET_OFFSET;
+
+public abstract class OldLogEntry extends AbstractLogEntry implements LogRecord {
+
+    public abstract Record record();
+
+    @Override
+    public long lastOffset() {
+        return offset();
+    }
+
+    @Override
+    public boolean isValid() {
+        return record().isValid();
+    }
+
+    @Override
+    public void ensureValid() {
+        record().ensureValid();
+    }
+
+    @Override
+    public int keySize() {
+        return record().keySize();
+    }
+
+    @Override
+    public boolean hasKey() {
+        return record().hasKey();
+    }
+
+    @Override
+    public ByteBuffer key() {
+        return record().key();
+    }
+
+    @Override
+    public int valueSize() {
+        return record().valueSize();
+    }
+
+    @Override
+    public boolean hasNullValue() {
+        return record().hasNullValue();
+    }
+
+    @Override
+    public ByteBuffer value() {
+        return record().value();
+    }
+
+    @Override
+    public boolean hasMagic(byte magic) {
+        return magic == record().magic();
+    }
+
+    @Override
+    public boolean hasTimestampType(TimestampType timestampType) {
+        return record().timestampType() == timestampType;
+    }
+
+    @Override
+    public long checksum() {
+        return record().checksum();
+    }
+
+    @Override
+    public long timestamp() {
+        return record().timestamp();
+    }
+
+    @Override
+    public TimestampType timestampType() {
+        return record().timestampType();
+    }
+
+    @Override
+    public long baseOffset() {
+        return iterator().next().offset();
+    }
+
+    @Override
+    public byte magic() {
+        return record().magic();
+    }
+
+    @Override
+    public CompressionType compressionType() {
+        return record().compressionType();
+    }
+
+    @Override
+    public long sequence() {
+        return 0;
+    }
+
+    @Override
+    public int sizeInBytes() {
+        return record().sizeInBytes() + LOG_OVERHEAD;
+    }
+
+    @Override
+    public String toString() {
+        return "OldLogEntry(" + offset() + ", " + record() + ")";
+    }
+
+    @Override
+    public void writeTo(ByteBuffer buffer) {
+        writeHeader(buffer, offset(), record().sizeInBytes());
+        buffer.put(record().buffer().duplicate());
+    }
+
+    @Override
+    public long pid() {
```

Comment:
For the following four functions: I'm wondering if 0s are good dummy values to use here, as 0s treated as meaningful values, or should we just throw an unsupportedoperationexception instead and rely on all callers to only trigger them if they are `EosLogEntry`?